cupy/fft/fft.py

```diff
@@ -119,21 +138,296 @@ def _fft(a, s, axes, norm, direction, value_type='C2C'):
+            raise ValueError("CUFFT plan dtype mismatch.")
+        # TODO: also check the strides and axes of the plan?
+
+    if overwrite_x and value_type == 'C2C':
+        out = a
+    elif out is None:
+        out = plan.get_output_array(a, order=order)
+    else:
+        plan.check_output_array(a, out)
+    plan.fft(a, out, direction)
+
+    # normalize by the product of the shape along the transformed axes
+    sz = np.prod([out.shape[ax] for ax in axes])
+    if norm is None:
+        if direction == cufft.CUFFT_INVERSE:
+            out /= sz
+    else:
+        out /= sqrt(sz)
+
+    return out
+
+
+def _fftn(a, s, axes, norm, direction, value_type='C2C', order='A', plan=None,
+          overwrite_x=False, out=None):
+    if norm not in (None, 'ortho'):
+        raise ValueError('Invalid norm value %s, should be None or \"ortho\".'
+                         % norm)
 
+    a = _convert_dtype(a, value_type)
+    if axes is None:
+        dim = a.ndim
+        axes = [i for i in six.moves.range(-dim, 0)]
+    axes = tuple(axes)
+
+    if (s is not None) and len(s) != len(axes):
+        raise ValueError("Shape and axes have different lengths.")
+
+    # sort the provided axes in ascending order
+    axes = tuple(sorted(np.mod(axes, a.ndim)))
+
+    if order == 'A':
+        if a.flags.f_contiguous:
+            order = 'F'
+        elif a.flags.c_contiguous:
+            order = 'C'
+        else:
+            a = cupy.ascontiguousarray(a)
+            order = 'C'
+    elif order not in ['C', 'F']:
+        raise ValueError("Unsupported order: {}".format(order))
+
+    a = _cook_shape(a, s, axes, value_type, order=order)
+    if order == 'C' and not a.flags.c_contiguous:
+        a = cupy.ascontiguousarray(a)
+    elif order == 'F' and not a.flags.f_contiguous:
+        a = cupy.asfortranarray(a)
+
+    a = _exec_fftn(a, direction, value_type, norm=norm, axes=axes,
+                   overwrite_x=overwrite_x, plan=plan, out=out)
     return a
 
 
+def _default_plan_type(a, s=None, axes=None):
+    """Determine whether to use separable 1d planning or nd planning."""
+    ndim = a.ndim
+    if ndim == 1 or not config.enable_nd_planning:
+        return '1d'
+
+    if axes is None:
+        if s is None:
+            dim = ndim
+        else:
+            dim = len(s)
+        axes = tuple([i % ndim for i in six.moves.range(-dim, 0)])
+    else:
+        # sort the provided axes in ascending order
+        axes = tuple(sorted([i % ndim for i in axes]))
+
+    if len(axes) == 1:
+        # use Plan1d to transform a  single axis
+        return '1d'
+    if len(axes) > 3 or not (np.all(np.diff(sorted(axes)) == 1)):
+        # PlanNd supports 1d, 2d or 3d transforms over contiguous axes
+        return '1d'
+    if (0 not in axes) and ((ndim - 1) not in axes):
+        # PlanNd only possible if the first or last axis is in axes.
+        return '1d'
+    return 'nd'
+
+
+def _default_fft_func(a, s=None, axes=None):
```

Comment:
The in-place operation is currently disabled because there's no `overwrite_x` argument here. Is this an oversight?

Reply:
Thanks, @grlee77, I can run successfully `fftn` in `cupyx.scipy.fftpack` with `overwrite_x=True`.